add re-estimation on nonce error, cleanup TxStore, misc fixes

cfal opened the pull request.

augustbleeds committed:
initial

augustbleeds committed:
add comment

augustbleeds committed:
retry on estimation failure if nonce error

augustbleeds committed:
check for nonce error in confirmer

augustbleeds committed:
fix txm tests

augustbleeds committed:
fix function invoke

augustbleeds committed:
fix err string handling

augustbleeds committed:
(debug) nonce (#369)

* debug logs

* publickey is for nonce manager, not contract address

* add skip validate to estimate

* change tag to pending

* make deep copy in txstore

* deep copy fix

* set x-apikey option for juno rpc

* fix bug

* add api key in few places

* fix debug err

* log

* tidy up logging

* remove nonce tracking from txstore

* clean up deep copy

* remove test

augustbleeds committed:
add delay to handling nonce

augustbleeds committed:
set apikey to empty string if none

augustbleeds committed:
set l2 instead of l1 gas info

augustbleeds committed:
add key

augustbleeds committed:
switch back to L1Gas

augustbleeds committed:
higher gas pad

augustbleeds committed:
remove chainid

augustbleeds committed:
remove invalidate flag + fix tests

augustbleeds committed:
simplify validate

augustbleeds committed:
add back skip validate flags for good

augustbleeds committed:
adjust token amount + add exp timer (#370)

* adjust token amount + add exp timer

* add back skip_validate flag

* modify config

augustbleeds committed:
bubble up sync error

augustbleeds committed:
verbose err logging

smickovskid committed:
Image bump

augustbleeds committed:
pass key not address into validate

augustbleeds committed:
dont error if no pending txs are found

cfal committed:
relayer/pkg/chainlink/txm/txm.go: print broadcast errors correctly

cfal committed:
relayer/pkg/chainlink/txm/txm.go: use smallest unconfirmed nonce for estimation

cfal committed:
relayer/pkg/chainlink/txm/txm.go: print fee estimation error

cfal committed:
relayer/pkg/chainlink/txm/txstore.go: add GetSmallestUnconfirmedNonce

augustbleeds commented on relayer/pkg/chainlink/txm/txm.go:
```diff
@@ -287,17 +287,22 @@ func (txm *starktxm) broadcast(ctx context.Context, publicKey *felt.Felt, accoun
 		Version:       starknetrpc.TransactionV3,
 		Signature:     []*felt.Felt{},
-		Nonce:         nonce,
+		Nonce:         unconfirmed,
```
will the mempool rejects this transaction if the nonce exists already in the mempool? (asking cause im not sure)

cfal replied:
we only use a nonce known to be unused when we submit (from nonce manager's NextSequence)

cfal committed:
relayer/pkg/starknet/client.go: dont create an account just to get nonce

cfal committed:
relayer/pkg/chainlink/txm/txm.go: add estimateFriFee

cfal committed:
TEMP: change handleNonceErr log to error

cfal committed:
relayer/pkg/chainlink/txm/txm.go: dont spin in the for loop when there are no items

cfal committed:
relayer/pkg/chainlink/txm/txm.go: use estimateFriFee

cfal committed:
relayer/pkg/chainlink/txm/txm.go: log feeder error

augustbleeds commented on relayer/pkg/chainlink/txm/txm.go:
```diff
@@ -171,18 +171,11 @@ func (txm *starktxm) broadcastLoop() {
-			}
-
-			// preserve tx queue: don't pull tx from queue until client is known to work
+		case tx := <-txm.queue:
```
nice catch

cfal committed:
integration-tests: revert tests to original nonce error branch

cfal committed:
relayer/pkg/chainlink/txm/txm.go: increase padding

sepolia's estimates seem to have strayed further and is now always
insufficient at 1.5x.

cfal committed:
relayer/pkg/chainlink/txm/txm.go: remove unused nonce code for now

cfal committed:
relayer/pkg/chainlink/txm/txm.go: change estimate fee startup message to Info, add error message when all attempts fail

cfal committed:
rename NewTestServer to NewTestFeederServer, fix lint error

archseer commented on relayer/pkg/chainlink/txm/nonce.go:
```diff
@@ -64,60 +64,72 @@ func (nm *nonceManager) HealthReport() map[string]error {
+	if err := nm.validate(publicKey); err != nil {
 		return nil, err
 	}
 
-	nm.lock.RLock()
-	defer nm.lock.RUnlock()
-	return nm.n[addr.String()][chainId], nil
+	return nm.n[publicKey.String()], nil
```
Minor nit, in both cases we validate which is a map lookup, then immediately fetch again. We could rename 'validate' into 'fetch' and have it return the value (we can ignore the value returned in 'Sync')

archseer commented on relayer/pkg/chainlink/txm/txstore.go:
```diff
@@ -109,6 +201,41 @@ func (c *ChainTxStore) Confirm(from *felt.Felt, hash string) error {
+	c.lock.RLock()
+	defer c.lock.RUnlock()
+
+	// empty slice if address isn't found
+	if err := c.validate(from); err != nil {
+		return nil
+	}
+
+	return c.store[from].GetSmallestUnconfirmedNonce()
```
Same with the validate implementation here

cfal committed:
relayer/pkg/chainlink/txm/txm.go: increase estimate attempts to 5

from ~300 testnet tx's, there was 1 case where we saw a 3rd attempt when
the nonce kept updating each iteration, so increase the number of
attempts

cfal committed:
relayer/pkg/chainlink/txm: rename ChainTxStore to AccountStore, remove wrapper TxStore functions

cfal committed:
relayer/pkg/chainlink/txm: remove NonceManager, combine with TxStore

cfal committed:
relayer/pkg/chainlink/txm/txm.go: cleanup logging and comments
